### DISCUSSION TITLE
kubevela always  restart

### DISCUSSION THREAD

**Discussion by xchkun** [kubevela/kubevela#2679] [Q&A]
docker logs contianer_name:
I1110 08:43:53.887738       1 traitdefinition_controller.go:62] "Reconcile traitDefinition" traitDefinition="vela-system/scaler"
I1110 08:43:53.892201       1 traitdefinition_controller.go:105] "Successfully update definitionRevision" definitionRevision="vela-system/scaler-v1"
I1110 08:43:53.944771       1 applicationcontext_controller.go:65] "Reconcile" applicationContext="vela-system/vela-env-default"
I1110 08:43:53.944957       1 application_controller.go:143] Build template
I1110 08:43:53.945013       1 application_controller.go:166] Apply application revision & component to the cluster
I1110 08:43:53.967752       1 componentdefinition_controller.go:112] "Successfully update definitionRevision" definitionRevision="vela-system/worker-v1"
I1110 08:43:53.968655       1 componentdefinition_controller.go:112] "Successfully update definitionRevision" definitionRevision="vela-system/webservice-v1"
I1110 08:43:53.968943       1 componentdefinition_controller.go:112] "Successfully update definitionRevision" definitionRevision="vela-system/task-v1"
I1110 08:43:53.969261       1 traitdefinition_controller.go:105] "Successfully update definitionRevision" definitionRevision="vela-system/sidecar-v1"
I1110 08:43:53.969532       1 traitdefinition_controller.go:105] "Successfully update definitionRevision" definitionRevision="default/flow-trait-cue-v1"
I1110 08:43:53.971502       1 apply.go:423] "Replace the existing appContext" appContext="vela-system/vela-env-default" revision it points to="vela-env-default-v1"
I1110 08:43:53.976815       1 application_controller.go:201] Check application health status
I1110 08:43:54.045228       1 applicationconfiguration.go:293] "ApplicationConfiguration" uid=8ed2fd3b-ba2f-4578-9559-dfff712a31b7 version=""
E1110 08:43:54.045358       1 runtime.go:78] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 724 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1d445a0, 0xc00f0d96e0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1d445a0, 0xc00f0d96e0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration.(*workloads).Apply(0xc00a44df50, 0x21edbe0, 0xc00920e5a0, 0x0, 0x0, 0x0, 0x31cf2c0, 0x0, 0x0, 0xc004b987f0, ...)
        /workspace/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/apply.go:106 +0xfee
github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration.(*OAMApplicationReconciler).ACReconcile(0xc00a460200, 0x21edbe0, 0xc00920e5a0, 0xc0079bfc00, 0xc00923fc00, 0x0)
        /workspace/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/applicationconfiguration.go:320 +0x11a8
github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationcontext.(*Reconciler).Reconcile(0xc001648000, 0xc00d88f8d0, 0xb, 0xc00d88f8a0, 0x10, 0xc00907f600, 0x0, 0x0, 0x0)
        /workspace/pkg/controller/core.oam.dev/v1alpha2/applicationcontext/applicationcontext_controller.go:109 +0x95e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000ccb40, 0x1c82400, 0xc00f23b780, 0x0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:235 +0x284
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000ccb40, 0x203000)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0000ccb40)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0038f4190)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0038f4190, 0x21a5160, 0xc00920e4e0, 0xc008973201, 0xc0001e1aa0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0038f4190, 0x3b9aca00, 0x0, 0x100000000000001, 0xc0001e1aa0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0038f4190, 0x3b9aca00, 0xc0001e1aa0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:170 +0x411
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 724 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1d445a0, 0xc00f0d96e0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration.(*workloads).Apply(0xc00a44df50, 0x21edbe0, 0xc00920e5a0, 0x0, 0x0, 0x0, 0x31cf2c0, 0x0, 0x0, 0xc004b987f0, ...)
        /workspace/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/apply.go:106 +0xfee
github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration.(*OAMApplicationReconciler).ACReconcile(0xc00a460200, 0x21edbe0, 0xc00920e5a0, 0xc0079bfc00, 0xc00923fc00, 0x0)
        /workspace/pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/applicationconfiguration.go:320 +0x11a8
github.com/oam-dev/kubevela/pkg/controller/core.oam.dev/v1alpha2/applicationcontext.(*Reconciler).Reconcile(0xc001648000, 0xc00d88f8d0, 0xb, 0xc00d88f8a0, 0x10, 0xc00907f600, 0x0, 0x0, 0x0)
        /workspace/pkg/controller/core.oam.dev/v1alpha2/applicationcontext/applicationcontext_controller.go:109 +0x95e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000ccb40, 0x1c82400, 0xc00f23b780, 0x0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:235 +0x284
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000ccb40, 0x203000)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0000ccb40)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0038f4190)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0038f4190, 0x21a5160, 0xc00920e4e0, 0xc008973201, 0xc0001e1aa0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0038f4190, 0x3b9aca00, 0x0, 0x100000000000001, 0xc0001e1aa0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0038f4190, 0x3b9aca00, 0xc0001e1aa0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:170 +0x411

and the uid=8ed2fd3b-ba2f-4578-9559-dfff712a31b7 is :
[root@node3 ~]# kubectl describe applicationcontext.core.oam.dev/vela-env-default -n vela-system
Name:         vela-env-default
Namespace:    vela-system
Labels:       app.oam.dev/app-revision-hash=63d1cf5be9a0b169
              cli.env.oam.dev/name=default
Annotations:  app.oam.dev/inplace-upgrade: true
              app.oam.dev/last-applied-configuration:
                {"kind":"Application","apiVersion":"core.oam.dev/v1beta1","metadata":{"name":"vela-env-default","namespace":"vela-system","creationTimesta...
API Version:  core.oam.dev/v1alpha2
Kind:         ApplicationContext
Metadata:
  Creation Timestamp:  2021-11-10T07:11:33Z
  Generation:          1
  Owner References:
    API Version:     core.oam.dev/v1beta1
    Controller:      true
    Kind:            Application
    Name:            vela-env-default
    UID:             d0900f32-4e6b-47c5-a4b0-aac077f70024
  Resource Version:  57496157
  Self Link:         /apis/core.oam.dev/v1alpha2/namespaces/vela-system/applicationcontexts/vela-env-default
  UID:               8ed2fd3b-ba2f-4578-9559-dfff712a31b7
Spec:
  Application Revision Name:  vela-env-default-v1
Events:                       <none>

thanks

**Comment by wangyikewxgm** [comment]
I think this a stale bug of vela 1.0.x. When user create an application without any component will cause this panic. Please upgrade your KubeVela to  latest version or make sure your apps  have at least one component.
